contracts/ETokensBundleVault.sol

```diff
@@ -0,0 +1,527 @@
+ * minUtilizationRate() limits, they will be allocated in the last one. Only if the last one doesn't accepts the funds,
+ * it will restart the allocation attempt starting from the first one. This makes the last and the first eToken more
+ * prone to receive funds when some of the eTokens don't accept deposits.
+ *
+ * When doing a withdrawal, if some of the funds can't be withdrawn from some of the eTokens (because of
+ * totalWithdrawable() validation), they will be withdrawn from the last one. Only if the last one doesn't accepts more
+ * withdrawals, it will restart the withdrawal attempt starting from the first one. This way, the last and the first
+ * eToken are more prone to have withdrawals when some of the eTokens don't accept withdrawals.
+ *
+ * @custom:security-contact security@ensuro.co
+ * @author Ensuro
+ */
+contract ETokensBundleVault is AccessControlUpgradeable, UUPSUpgradeable, ERC4626Upgradeable {
+  using SafeCast for uint256;
+  using WadRayMath for uint256;
+
+  uint256 private constant HUNDRED_PERCENT = 1e18;
+
+  bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+  bytes32 public constant REBALANCER_ROLE = keccak256("REBALANCER_ROLE");
+  bytes32 public constant CHANGE_PERCENTAGE_ROLE = keccak256("CHANGE_PERCENTAGE_ROLE");
+  bytes32 public constant REORDER_ROLE = keccak256("REORDER_ROLE");
+  bytes32 public constant GUARDIAN_ROLE = keccak256("GUARDIAN_ROLE");
+
+  struct Underlying {
+    EToken etk;
+    uint16 percentage; // percentage stored with 4 decimals
+  }
+
+  Underlying[] internal _underlying;
+
+  event UnderlyingChanged(EToken etk, uint256 index, uint256 percentage);
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() {
+    _disableInitializers();
+  }
+
+  /**
+   * @dev Initializes the ERC4626CashFlowLender
+   */
+  function initialize(
+    EToken[] calldata etks,
+    uint256[] calldata percentages
+  ) public virtual initializer {
+    __ETokensBundleVault_init(etks, percentages);
+  }
+
+  // solhint-disable-next-line func-name-mixedcase
+  function __ETokensBundleVault_init(
+    EToken[] calldata etks,
+    uint256[] calldata percentages
+  ) internal onlyInitializing {
+    __UUPSUpgradeable_init();
+    __AccessControl_init();
+    require(etks.length > 0, "ETokensBundleVault: the vault must have always at least one ETK");
+    __ERC4626_init(IERC20Upgradeable(address(etks[0].policyPool().currency())));
+    __ETokensBundleVault_init_unchained(etks, percentages);
+  }
+
+  // solhint-disable-next-line func-name-mixedcase
+  function __ETokensBundleVault_init_unchained(
+    EToken[] calldata etks,
+    uint256[] calldata percentages
+  ) internal onlyInitializing {
+    require(
+      etks.length == percentages.length,
+      "ETokensBundleVault: etks and percentages lengths differ"
+    );
+    _setupRole(DEFAULT_ADMIN_ROLE, _msgSender());
+
+    IPolicyPool pool;
+    uint256 totalPercentage;
+    for (uint256 i; i < etks.length; i++) {
+      if (i == 0) {
```

Comment:
Aca no podrias hacer: 

```js
IPolicyPool pool = etks[0]
uint256 totalPercentage;
for (uint256 i; i < etks.length; i++) {
        require(
          pool == etks[i].policyPool(),
          "ETokensBundleVault: Can't mix eTokens from different PolicyPool"
        );
}
```

y sacamos el if

Reply:
No, etks no podría estar vacío. 

Pero haciendo eso que ponés hace la llamada a `policyPool()` una vez más. Ya que lo llama afuera del for y en la primer iteración. No es que importe mucho en términos de gas, pero lo dejaría así.